docs/dev/pagination-v2.md

```diff
@@ -0,0 +1,175 @@
+
+# REST API
+## Initial Query
+```
+POST /_plugins/_sql
+{
+    "query" : "...",
+    "fetch_size": N
+}
+```
+
+Response
+```
+{
+  "cursor": /* cursor_id */,
+  "datarows": [
+    // ...
+    ],
+  "schema" : [
+    // ...
+  ]
+}
+```
+
+## Get Page
+```
+POST /_plugins/_sql
+{
+  "cursor": "<cursor_id>"
+}
+```
+response is the same as initial response if this is not the last page.
+
+If this is the last page, the `cursor` property is ommitted. The cursor is closed automatically.
+
+## Close Cursor
+```
+POST /_plugins/_sql/close
+{
+    "cursor" : "<cursor_id>"
+}
+```
+
+Response
+```json
+{
+    "succeeded": true
+}
+
+```
+v1 engine always returns success on close even if the cursor is no longer alive as long as it recognizes cursor_id as a valid cursor identifier.
+
+v1 engine responds with an error such this if cursor_id is not a valid cursor identifier.
+```json5
+{
+    "error": {
+        // ...
+    },
+    "status": 400
+}
+```
+## Communication with OpenSearch
+
+There are several options to communicate with OpenSearch. Depending on the type of query, it may make sense to issue different requests to OpenSearch.
+
+
+### Simple Requests
+
+For some queries, SQL plugin can fetch all data from OpenSearch at once.
+
+```mermaid
+sequenceDiagram
+  participant sql_client
+  participant sql_plugin
+  participant opensearch
+  
+  sql_client ->> sql_plugin: query with fetch_size
+  sql_plugin -->> opensearch: DSL query
+  opensearch -->> sql_plugin: complete result set
+  sql_plugin -->> sql_client: subset of results, cursor_id
+  loop while more data
+    sql_client -->> sql_plugin: request a subset of results for cursor_id
+    sql_plugin -->> sql_client:  subset of results, new cursor_id
+  end
+  sql_client ->> sql_plugin: close cursor_id
+  sql_plugin ->> sql_client: ACK close cursor
+```
+
+### Scroll-backed Requests
+
+```mermaid
+sequenceDiagram
+  participant sql_client
+  participant sql_plugin
+  participant opensearch
+  
+  sql_client ->> sql_plugin: query with fetch_size
+  sql_plugin -->> opensearch: scroll query
+  opensearch -->> sql_plugin: subset of results and scroll_id
+  sql_plugin -->> sql_client: subset of results and cursor_id
+  loop while more data
+    sql_client -->> sql_plugin: request a subset of results for cursor_id
+    sql_plugin -->> opensearch: request a subset of results for scroll_id
+    opensearch -->> sql_plugin: subset of results, new scroll_id
+    sql_plugin -->> sql_client: subset of results, new cursor_id
+  end
+  sql_client -->> sql_plugin: close cursor_id
+  sql_plugin -->> opensearch: close scroll_id
+  opensearch -->> sql_plugin: ACK close cursor
+  sql_plugin ->> sql_client: ACK close cursor
+```
+
+
+Scroll requests are resource intensive and are [not recommended](https://opensearch.org/docs/latest/api-reference/scroll/) for frequent user queries.
+
+Search after is a simpler API that can be used in these scenarios. 
+
+It does require query to use `ORDER BY` clause.
```

Comment:
Sorry, what is "Search after"?
Why it require `ORDER BY` clause?

Reply:
[Search after](https://opensearch.org/docs/2.2/opensearch/search/paginate/#the-search_after-parameter) is one way to paginate results in OpenSearch but to use it the OpenSearch query needs to include a sort clause.

Technically we can pick a field to search on and use Search After API even for SQL queries that do not have an `ORDER BY` clause.